stratosphere/ams_mitm/source/set_mitm/setsys_mitm_service.cpp

```diff
@@ -87,6 +88,87 @@ namespace ams::mitm::settings {
+            /* Open external database file. */
+            fs::FileHandle file;
+            R_TRY(fs::OpenFile(std::addressof(file), g_external_bluetooth_db_location, fs::OpenMode_Read));
+            ON_SCOPE_EXIT { fs::CloseFile(file); };
+
+            /* Read number of database entries stored in external database. */
+            size_t total_entries;
+            R_TRY(fs::ReadFile(file, 0, std::addressof(total_entries), sizeof(total_entries)));
+
+            u64 db_offset = sizeof(total_entries);
+            if (total_entries > db_max_size) {
+                /* Cap number of database entries read to size of database on this firmware */
+                total_entries = db_max_size;
+
+                /* Pairings are stored from least to most recent. Add offset to skip the older entries that won't fit */
+                db_offset += (total_entries - db_max_size) * sizeof(SetSysBluetoothDevicesSettings);
+            }
+
+            /* Read database entries. */
+            R_TRY(fs::ReadFile(file, db_offset, db, total_entries * sizeof(SetSysBluetoothDevicesSettings)));
+
+            /* Convert to old database format if running on a firmware below 13.0.0 */
+            if (hos::GetVersion() < hos::Version_13_0_0) {
+                for (u32 i = 0; i < total_entries; ++i) {
+                    util::SNPrintf(db[i].name.name, sizeof(db[i].name), "%s", db[i].name2);
+                    //std::memcpy(std::addressof(db[i].name), db[i].name2, sizeof(db[i].name));
+                    std::memset(db[i].name2, 0, sizeof(db[i].name2));
+                }
+            }
+
+            *entries_read = total_entries;
+
+            return ResultSuccess();
+        }
+
+        Result WriteExternalBluetoothDatabase(const SetSysBluetoothDevicesSettings *db, size_t total_entries) {
+            /* Open external database file. */
+            fs::FileHandle file;
+            R_TRY(fs::OpenFile(std::addressof(file), g_external_bluetooth_db_location, fs::OpenMode_Write | fs::OpenMode_AllowAppend));
+            ON_SCOPE_EXIT { fs::CloseFile(file); };
+
+            /* Write number of database entries. */
+            R_TRY(fs::WriteFile(file, 0, std::addressof(total_entries), sizeof(total_entries), fs::WriteOption::None));
+
+            /* Write database entries. */
+            u64 db_offset = sizeof(total_entries);
+            if (hos::GetVersion() < hos::Version_13_0_0) {
+                /* Convert to new database format if running on a firmware below 13.0.0 */
+                auto tmp_entry = std::make_unique<SetSysBluetoothDevicesSettings>();
+                for (u32 i = 0; i < total_entries; ++i) {
+                    std::memcpy(tmp_entry.get(), std::addressof(db[i]), sizeof(SetSysBluetoothDevicesSettings));
+                    //std::memcpy(tmp_entry->name2, std::addressof(tmp_entry->name), sizeof(tmp_entry->name));
```

Comment:
Not sure why this is commented out.

Also, this whole scope is really dense, add spacing and ideally comments.

Reply:
Oops, replaced the memcpy call with with SNPrintf, forgot to remove.